core/arch/arm/plat-imx/drivers/imx_caam.c

```diff
@@ -9,9 +9,148 @@
+
+static void caam_enable_clocks(void)
+{
+	vaddr_t ccm_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);
+
+	io_setbits32(ccm_base + CCM_CCGR0,
+		     CCM_CCGR0_CAAM_WRAPPER_IPG  |
+		     CCM_CCGR0_CAAM_WRAPPER_ACLK |
+		     CCM_CCGR0_CAAM_SECURE_MEM);
+
+	if (soc_is_imx6dqp() || soc_is_imx6sdl() || soc_is_imx6dq())
+		io_setbits32(ccm_base + CCM_CCGR6,
+			     CCM_CCGR6_EMI_SLOW);
+}
+
+static TEE_Result imx_caam_reset_jr(struct imx_caam_ctrl *ctrl)
+{
+	uint32_t reg_val = 0;
+	uint32_t timeout = 1000;
+
+	io_write32((vaddr_t)&ctrl->jrcfg[MKVB_JR].jrcr, 1);
+	do {
+		reg_val = io_read32((vaddr_t)&ctrl->jrcfg[MKVB_JR].jrintr);
+		reg_val &= 0xc;
+	} while ((reg_val == 0x4) && --timeout);
+
+	if (!timeout)
+		return TEE_ERROR_SECURITY;
+
+	io_write32((vaddr_t)&ctrl->jrcfg[MKVB_JR].jrcr, 1);
+	do {
+		reg_val = io_read32((vaddr_t)&ctrl->jrcfg[MKVB_JR].jrintr);
+		reg_val &= 0xc;
+	} while ((reg_val & 0x1) && --timeout);
+
+	if (!timeout)
+		return TEE_ERROR_SECURITY;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mkvb_init_jr(struct imx_mkvb *mkvb)
+{
+	struct imx_caam_ctrl *ctrl = mkvb->ctrl;
+	TEE_Result ret = TEE_ERROR_SECURITY;
+
+	ret = imx_caam_reset_jr(ctrl);
+	if (ret)
+		return ret;
+
+	mkvb->njobs = 4;
+	io_write32((vaddr_t)&ctrl->jrstartr, MKVB_JR1_START);
+	io_write32((vaddr_t)&ctrl->jrcfg[MKVB_JR].irbar_ls,
+		   virt_to_phys(&mkvb->jr.inring));
+	io_write32((vaddr_t)&ctrl->jrcfg[MKVB_JR].irsr, mkvb->njobs);
+	io_write32((vaddr_t)&ctrl->jrcfg[MKVB_JR].orbar_ls,
+		   virt_to_phys(&mkvb->jr.outring));
+	io_write32((vaddr_t)&ctrl->jrcfg[MKVB_JR].orsr, mkvb->njobs);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result caam_get_mkvb(uint8_t *dest)
+{
+	struct imx_mkvb mkvb = { 0 };
+	TEE_Result ret = TEE_ERROR_SECURITY;
+	paddr_t desc_paddr = 0;
+	int counter = 0;
+
+	mkvb.ctrl = (struct imx_caam_ctrl *)
+		core_mmu_get_va(CAAM_BASE, MEM_AREA_IO_SEC);
+
+	ret = mkvb_init_jr(&mkvb);
+	if (ret)
+		goto out;
+
+	mkvb.descriptor[0] = MKVB_DESC_HEADER;
+	mkvb.descriptor[1] = MKVB_DESC_SEQ_OUT;
+	mkvb.descriptor[2] = virt_to_phys(&mkvb.outbuf);
+	cache_operation(TEE_CACHEFLUSH, &mkvb.outbuf, MKVB_SIZE);
+	mkvb.descriptor[3] = MKVB_DESC_BLOB;
+	cache_operation(TEE_CACHEFLUSH, &mkvb.descriptor,
+			sizeof(mkvb.descriptor));
+
+	desc_paddr = virt_to_phys(&mkvb.descriptor);
+	memcpy(&mkvb.jr.inring, &desc_paddr, sizeof(desc_paddr));
+	cache_operation(TEE_CACHEFLUSH, &mkvb.jr,
+			sizeof(mkvb.jr.inring[0]));
+
+	/* Tell CAAM that one job is available */
+	io_write32((vaddr_t)&mkvb.ctrl->jrcfg[MKVB_JR].irjar, 1);
+
+	/* Busy loop until job is completed */
+	while (io_read32((vaddr_t)&mkvb.ctrl->jrcfg[MKVB_JR].orsfr) != 1) {
+		counter++;
+		if (counter > 10000)
+			goto out;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, &mkvb.jr, sizeof(mkvb.jr));
+	cache_operation(TEE_CACHEINVALIDATE, &mkvb.outbuf, MKVB_SIZE);
+	DHEXDUMP(&mkvb.outbuf, MKVB_SIZE);
+
+	if (mkvb.jr.outring[0].status != 0)
+		ret = TEE_ERROR_SECURITY;
+		goto out;
+
```

Comment:
Missing { } otherwise it always ends up going to 'goto out'.

Reply:
Yes, this is already addressed in my local branch. I plan to rebase onto the CAAM work, but that might take a bit of refactoring to correctly disable the CAAM driver after HUK generation.

Reply:
Yes, this is already addressed in my local branch. I plan to rebase onto the CAAM work, but that might take a bit of refactoring to correctly disable the CAAM driver after HUK generation.

Reply:
Yes, this is already addressed in my local branch. I plan to rebase onto the CAAM work, but that might take a bit of refactoring to correctly disable the CAAM driver after HUK generation.

Reply:
Great, thanks for the heads up.